---
title: "Section 2: Bioconductor basics with GRanges and Biostrings"
author: "Aditya Rao"
date: "2022-12-01"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
my_working_dir = "/Users/adityarao/personal_projects/online_courses/HarvardX Data Analysis for Genomics Professional Certificate/HarvardX PH525.5x Introduction to Bioconductor Annotation and Analysis of Genomes and Genomic Assays"
setwd(my_working_dir)
library(data.table)
library(ggplot2)
```


```{r}
library(ERBS)
data(HepG2)

class(HepG2)
```

```{r}
# install ERBS
# install_github("genomicsclass/ERBS")

# load GM12878 and HepG2 objects from ERBS package
library(ERBS)
data(GM12878)
data(HepG2)

# inspect HepG2 GRanges object
class(HepG2)
HepG2
values(HepG2)

# seqnames extracts chromosome names
seqnames(HepG2)     # stored as type Rle
chr = seqnames(HepG2)
as.character(chr)    # view as character type

# make a table of numbers of sequences on each chromosome
table(chr)
table(chr)[1:24]    # restrict to autosomes, X and Y

# GRanges can be subsetted and ordered
HepG2[chr=="chr20",]
x = HepG2[order(HepG2),]
seqnames(x)     # demonstrate usefulness of Rle type
as.character(seqnames(x))

```

```{r}
summary(mcols(HepG2)$signalValue)
HepG2[mcols(HepG2)$signalValue == max(mcols(HepG2)$signalValue)]
table(seqnames(HepG2))
summary(width(ranges(HepG2)))
```

```{r}
hist(width(ranges(HepG2)))
```

```{r}
library(IRanges)

ran = IRanges(101,200)
ran*2

narrow(ran, start=20)

ran+25
```

```{r}
ran2 = IRanges(c(1,11,21),c(3,15,27))
ran2
sum(width(ran2))

ran3 = IRanges(c(101,106,201,211,221,301,306,311,351,361,401,411,501),c(150,160,210,270,225,310,310,330,390,380,415,470,510))
library(ph525x)
plotRanges(ran3)
```

```{r}
range(ran3)
reduce(ran3)
gaps(ran3)
sum(width(gaps(ran3)))
```

```{r}
disjoin(ran3)
```


```{r}
plotRanges(disjoin(ran3))
```

```{r}
par(mfrow=c(2,1))
plotRanges(ran3,xlim=c(0,600))
plotRanges(resize(ran3,1),xlim=c(0,600))
```

```{r}
x = GRanges("chr1", IRanges(c(1,101),c(50,150)), strand=c("+","-"))
x
```

```{r}
plotGRanges = function(x) plotRanges(ranges(x))
par(mfrow=c(2,1))
plotGRanges(x)
plotGRanges(resize(x,1))
```

```{r}
x = GRanges("chr1", IRanges(c(101,201,401,501),c(150,250,450,550)), strand="+")
y = GRanges("chr1", IRanges(c(101,221,301,401,541),c(150,250,350,470,550)), strand="+")

GRangesList(x,y)
```

```{r}
c(x,y)
```

```{r}
disjoin(c(x,y))[disjoin(c(x,y)) %over% x & disjoin(c(x,y)) %over% y]
```

```{r}
sum(width(ranges(disjoin(c(x,y))[disjoin(c(x,y)) %over% x & disjoin(c(x,y)) %over% y])))
```

```{r}
sum(width(ranges(disjoin(c(x,y))[!(disjoin(c(x,y)) %over% x & disjoin(c(x,y)) %over% y)])))
```

```{r}
z = GRanges("chr1", range(ranges(x)), strand=c("-"))
x[x %over% z]
```

```{r}
ranges(HepG2)[17]
distanceToNearest(HepG2[17],GM12878)
```

```{r}
GM12878[945]
```

```{r}
distanceToNearest(HepG2,GM12878)
```

```{r}
sum(mcols(distanceToNearest(HepG2,GM12878))$distance < 2000)/length(HepG2)
```

```{r}
library(Homo.sapiens)
ghs = genes(Homo.sapiens)
```

```{r}
ghs
genome(ghs)
table(seqnames(ghs))
hist(width(ranges(ghs)))
median(width(ranges(ghs)))
```

```{r}
res = findOverlaps(HepG2,GM12878)
erbs = HepG2[queryHits(res)]
erbs = granges(erbs)
erbs
```

```{r}
erbs2= intersect(HepG2,GM12878)
erbs2
```

```{r}
erbs %over% erbs2
erbs2 %over% erbs
```

```{r}
ghs = genes(Homo.sapiens)
tssghs = resize(ghs,1)
tssghs["100113402"]
```

```{r}
tssghs[nearest(erbs[4],tssghs)]
```

```{r}
select(Homo.sapiens,keys=2101,keytype = "GENEID",columns="SYMBOL")
```

```{r}
library(Biostrings)

# basics of DNAStrings
dna <- DNAString("TCGAGCAAT")    # define a DNAString
dna
length(dna)    # number of bases in a DNAString
DNAString("JQX")    # error - invalid bases
DNAString("NNNACGCGC-TTA-CGGGCTANN")    # valid sequence with unknowns and gaps
dna[4:6]    # extract a substring
as.character(dna)    # convert DNAString to character
```

```{r}
# basics of DNAStringSets
set1 <- DNAStringSet(c("TCA", "AAATCG", "ACGTGCCTA", "CGCGCA", "GTT", "TCA"))    # define a DNAStringSet
set1
set1[2:3]    # extract subset of sequences
set1[[4]]    # extract one sequence as a single DNAString
length(set1)    # number of DNAstrings in set
width(set1)    # size of each DNAString
duplicated(set1)    # detect which sequences are duplicated
unique(set1)    # keep only unique sequences
sort(set1)
```

```{r}
dna_seq <- DNAString("ATCGCGCGCGGCTCTTTTAAAAAAACGCTACTACCATGTGTGTCTATC")

# analyze DNAStrings
letterFrequency(dna_seq, "A")    # count A in sequence
letterFrequency(dna_seq, "GC")    # count G or C in sequence
dinucleotideFrequency(dna_seq)    # frequencies of all dinucleotides
trinucleotideFrequency(dna_seq)    # frequencies of all trinucleotides

# convert DNAStrings
reverseComplement(dna_seq)    # find reverse complement
translate(dna_seq)    # amino acid translation
```

```{r}
# count and match on individual Biostrings
dna_seq <- DNAString("ATCGCGCGCGGCTCTTTTAAAAAAACGCTACTACCATGTGTGTCTATC")
dna_seq
countPattern("CG", dna_seq)    # pattern "CG" occurs 5 times
matchPattern("CG", dna_seq)    # locations of pattern "CG"
start(matchPattern("CG", dna_seq))    # start locations of the pattern
matchPattern("CTCTTTTAAAAAAACGCTACTACCATGTGT", dna_seq)    # match patterns of any length

# check for pattern and its reverse complement
countPattern("TAG", dna_seq)
countPattern(reverseComplement(DNAString("TAG")), dna_seq)

# count and match on sets of Biostrings
set2 <- DNAStringSet(c("AACCGGTTTCGA", "CATGCTGCTACA", "CGATCGCGCCGG", "TACAACCGTACA"))
set2
vcountPattern("CG", set2)    # CG counts for entire DNAStringSet
vmatchPattern("CG", set2)
vmatchPattern("CG", set2)[[1]]    # access matches for the first element of the DNAStringSet
```

```{r}
eco <- DNAString("GGTTTCACCGCCGGTAATGAAAAAGGCGAACTGGTGGTGCTTGGACGCAACGGTTCCGACTACTCTGCTGCGGTGCTGGCTGCCTGTTTACGCGCCGATTGTTGCGAGATTTGGACGGACGTTGACGGGGTCTATACCTGCGACCCGCGTCAGGTGCCCGATGCGAGGTTGTTGAAGTCGA")
eco
```

```{r}
length(eco)
countPattern("ATG", eco)
trinucleotideFrequency(eco)
matchPattern("ATG", eco)
aastr = translate(eco[17:length(eco)])
aastr
matchPattern("*",aastr)
aastr2 = aastr[1:52]
aastr2
letterFrequency(aastr2,"HKR")
letterFrequency(aastr2,"DE")
```

```{r}
library(ERBS)
library(GenomicRanges)
data(HepG2)
data(GM12878)
res = findOverlaps(HepG2,GM12878)
erbs = HepG2[queryHits(res)]
erbs = granges(erbs)
```

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
erbs_seq = getSeq(Hsapiens,erbs)
erbs_seq
```

```{r}
vec = rep(0,length(erbs_seq))
for(i in 1:length(erbs_seq)){
  vec[i] = letterFrequency(erbs_seq[[i]],"GC")/length(erbs_seq[[i]])
}
median(vec)
```

```{r}
erbs_seq2 = getSeq(Hsapiens,shift(erbs,1000))
vec2 = rep(0,length(erbs_seq2))
for(i in 1:length(erbs_seq2)){
  vec2[i] = letterFrequency(erbs_seq2[[i]],"GC")/length(erbs_seq2[[i]])
}
median(vec2)
```

```{r}
library(Homo.sapiens)
g <- genes(Homo.sapiens)
g
```

```{r}
# restrict to genes on chr21
chr21 <- keepSeqlevels(g, "chr21", pruning.mode = "coarse")
chr21
max(width(ranges(chr21)))
sum(as.character(strand(chr21)) == "+")/length(as.character(strand(chr21)))
```

```{r}
chr21_seq = getSeq(Hsapiens,chr21)
vec = rep(0,length(chr21_seq))
for(i in 1:length(chr21_seq)){
  vec[i] = letterFrequency(chr21_seq[[i]],"GC")/length(chr21_seq[[i]])
}
median(vec)


```

```{r}
countPattern("ATG",chr21_seq[[5]])
matchPattern("ATG",chr21_seq[[5]])
translate(chr21_seq[[5]][32:length(chr21_seq[[5]])])
```

```{r}
snca <- genes(Homo.sapiens, filter=list(GENEID="6622"))
snca
snca_seq = getSeq(Hsapiens,snca)
vmatchPattern("ACTGTGAA",snca_seq)
vmatchPattern("ACTGTGAA",reverseComplement(snca_seq))
```

```{r}
# Load the GRanges corresponding to human genes from the Homo.sapiens package
g <- genes(Homo.sapiens)

# Load the ESRRA binding site GRanges in GM12878 cells
data(GM12878)

g_seq = getSeq(Hsapiens,g)
g_promseq = getSeq(Hsapiens,shift(g,2000))
sort(shift(g,2000))
shift(sort(g),2000)[100]
sort(promoters(g,upstream=2000))[100]
prom = flank(g, 2000)
prom[100]
```

```{r}
over = findOverlaps(GM12878,prom)
over
GM12878[queryHits(over)]
prom[subjectHits(over)]

length(unique(ranges(GM12878[queryHits(over)])))
length(unique(ranges(prom[subjectHits(over)])))

over2 = findOverlaps(prom,GM12878)
length(unique(prom[queryHits(over2)]))
length(unique(g[queryHits(over2)]))
```

```{r}
rprom = shift(prom, 10000)
over2 = findOverlaps(GM12878,rprom)
over2
length(unique(GM12878[queryHits(over2)]))
757/117
```


















